server/src/main/java/org/elasticsearch/indices/recovery/RemoteRecoveryTargetHandler.java

```diff
@@ -133,10 +136,10 @@ public void finalizeRecovery(final long globalCheckpoint, final long trimAboveSe
     @Override
     public void handoffPrimaryContext(final ReplicationTracker.PrimaryContext primaryContext) {
         TransportFuture<TransportResponse.Empty> handler = new TransportFuture<>(EmptyTransportResponseHandler.INSTANCE_SAME);
-        transportService.sendRequest(
-            targetNode, PeerRecoveryTargetService.Actions.HANDOFF_PRIMARY_CONTEXT,
-            new RecoveryHandoffPrimaryContextRequest(recoveryId, shardId, primaryContext),
-            TransportRequestOptions.builder().withTimeout(recoverySettings.internalActionTimeout()).build(), handler);
+        final TransportRequestOptions requestOptions = waitIndefinitelyOnRelocation ? TransportRequestOptions.EMPTY :
```

Comment:
Would be nice to eventually have https://github.com/elastic/elasticsearch/pull/64098 for this if we're ok with hooking into this request so that we don't waste a generic thread here for a non-trivial amount of time.